From 0b7ac6dcde7b251dff52c0f39c576e2b75f46db5 Mon Sep 17 00:00:00 2001
From: Serhiy Boiko <serhiy.boiko@plvision.eu>
Date: Wed, 2 Feb 2022 18:32:27 +0200
Subject: [PATCH 40/40] Fix graft error message

Add ets graft handler to prevent error message when creating
tbf qdisc.

JIRA: SWITCHDEV-000
Signed-off-by: Serhiy Boiko <serhiy.boiko@plvision.eu>
Change-Id: Iff86eb8a37857de4f9ec504fa35b5de4fbf5195c
---
 .../net/ethernet/marvell/prestera/prestera_qdisc.c  | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/net/ethernet/marvell/prestera/prestera_qdisc.c b/drivers/net/ethernet/marvell/prestera/prestera_qdisc.c
index 79a42bf6c1f8..6c66dc07cdd4 100644
--- a/drivers/net/ethernet/marvell/prestera/prestera_qdisc.c
+++ b/drivers/net/ethernet/marvell/prestera/prestera_qdisc.c
@@ -70,6 +70,16 @@ static void prestera_qdisc_root_destroy(struct prestera_port *port)
 	port->qdisc->root = 0;
 }
 
+static int prestera_qdisc_graft(struct prestera_port *port, u8 band,
+				u32 child_handle)
+{
+	if (band < IEEE_8021QAZ_MAX_TCS &&
+	    port->qdisc->tbf[band] == child_handle)
+		return 0;
+
+	return -EOPNOTSUPP;
+}
+
 static u32
 prestera_qdisc_tbf_burst(struct tc_tbf_qopt_offload_replace_params *p)
 {
@@ -142,6 +152,9 @@ static int __prestera_setup_tc_ets(struct prestera_port *port,
 		return 0;
 	case TC_ETS_STATS:
 		return 0;
+	case TC_ETS_GRAFT:
+		return prestera_qdisc_graft(port, p->graft_params.band,
+					    p->graft_params.child_handle);
 	default:
 		return -EOPNOTSUPP;
 	}
-- 
2.17.1

